abdiff/core/run_ab_transforms.py

```diff
@@ -168,47 +195,61 @@ def run_docker_container(
         f"Container '{container.id}' (Docker image: {docker_image}) "
         f"RUNNING transform for '{source}' input_file: {input_file}."
     )
-    return container
-
 
-def wait_for_containers(
-    containers: list[Container], timeout: int = 180
-) -> None | list[str]:
-    """Given a list of running containers, wait until all containers exit.
-
-    If there are any containers that exited with an error, a list of IDs
-    for failed containers is returned. For more information regarding the
-    specific cause of the error, consult the logs.
-    """
-    exited_containers: list[str] = []
-    failed_containers: list[str] = []
-    start_time = time.time()
-    while len(exited_containers) < len(containers):
-        running_containers = [
-            container for container in containers if container.id not in exited_containers
-        ]
-        if time.time() - start_time >= timeout:
-            raise DockerContainerRuntimeExceededTimeoutError(
-                containers=running_containers, timeout=timeout
-            )
-        for container in running_containers:
+    exception = None
+    try:
+        start_time = perf_counter()
+        while True:
+            time.sleep(0.5)
             container.reload()
             if container.status == "exited":
                 logger.info(f"Container {container.id} exited.")
-                exited_containers.append(container.id)  # type: ignore[arg-type]
-                if (exit_code := container.attrs["State"]["ExitCode"]) == 0:
-                    logger.info(f"Container {container.id} ran successfully.")
-                else:
-                    logger.error(
-                        f"Container {container.id} exited with an error: {exit_code}. "
-                        "Check logs for details."
-                    )
-                    failed_containers.append(container.id)  # type: ignore[arg-type]
-        time.sleep(0.25)
-
-    if any(failed_containers):
-        return failed_containers
-    return None
+                break
+
+            if time.perf_counter() - start_time > timeout:
+                logger.error(
+                    f"Container {container.id} timed out after {timeout} seconds"
+                )
+                container.stop()
+                exception = DockerContainerTimeoutError(
+                    container_id=container.id, timeout=timeout
+                )
+                break
+
+    except Exception as e:
+        exception = e  # type: ignore[assignment]
+        logger.exception("Unhandled exception while waiting for container to complete.")
```

Comment:
Hmm, curious what kind of exception would be handled by this? 🤔 Just to make sure we're on the same page, this would not capture the error that caused the container to fail--it is handling exceptions (if any) raised by `container.reload()` and `container.status`, right?

Reply:
_[Had this response saved from last Friday, forgot to send...]_

That's right!  [This line/test](https://github.com/MITLibraries/transmogrifier-ab-diff/pull/37/files#diff-ef381e309d69d9280a44bcd8e1f647b38878fd2167b92011767f6f351ce4363dR246) mocks just that.  If for any reason the `Container.reload()` method fails we _should_ still get both a `Container` and `Exception` returned from `run_docker_container()` such that we can aggregate logs.

One might argue even [this code](https://github.com/MITLibraries/transmogrifier-ab-diff/blob/TIMX-356-limit-parallel-containers/abdiff/core/run_ab_transforms.py#L174-L193) that invokes the Docker container in a detatched state would be inside the `try/except` block... But if that fails, then something is wrong like no Docker environment, or something with the configurations, and I think it's okay that an exception would bubble and kill the `run_ab_transforms()` call immediately.